packages/opentelemetry-exporter-datadog/src/transform.ts

```diff
@@ -0,0 +1,299 @@
+  if (span.parentSpanId === undefined) {
+    const origin = createOriginString(span);
+    if (origin) {
+      ddSpanBase.setTag(DD_ORIGIN, origin);
+    }
+    if (version) {
+      ddSpanBase.setTag(VERSION_KEY, version);
+    }
+  }
+
+  // set span attibutes as tags - takes precedence over env vars
+  // span tags should be strings
+  for (const [key, value] of Object.entries(span.attributes)) {
+    if (value !== undefined) {
+      ddSpanBase.setTag(key, value);
+    }
+  }
+
+  // filter for internal requests to trace-agent and set sampling rate
+  const samplingRate = getSamplingRate(span);
+  const internalRequest = isInternalRequest(span);
+
+  if (internalRequest) {
+    ddSpanBase.setTag(SAMPLE_RATE_METRIC_KEY, USER_REJECT);
+  } else if (samplingRate !== undefined) {
+    ddSpanBase.setTag(SAMPLE_RATE_METRIC_KEY, samplingRate);
+  }
+
+  // set span duration and mark as finished span so that it is exported
+  const duration =
+    hrTimeToMilliseconds(span.endTime) - hrTimeToMilliseconds(span.startTime);
+  ddSpanBase._duration = duration;
+  ddSpanBaseContext._isFinished = true;
+
+  return ddSpanBase;
+}
+
+function getTraceContext(span: ReadableSpan): typeof Span[] {
+  return [
+    id(span.spanContext['traceId']),
+    id(span.spanContext['spanId']),
+    span.parentSpanId ? id(span.parentSpanId) : null,
+  ];
+}
+
+function createDefaultTags(tags: string | undefined): object {
+  // Parse a string of tags typically provided via environment variables.
+  // The expected string is of the form: "key1:value1,key2:value2"
+
+  const tagMap = tags
+    ? tags
+        .split(',')
+        .reduce((tags: { [key: string]: string }, kvPair: string) => {
+          const kvTuple = kvPair.split(':');
+          tags[kvTuple[0]] = kvTuple[1];
+          return tags;
+        }, {})
+    : {};
+
+  // ensure default tag env var or  arg is not malformed
+  if (
+    Object.keys(tagMap).indexOf('') >= 0 ||
+    Object.values(tagMap).indexOf('') >= 0 ||
+    Object.values(tagMap).some((v: string) => {
+      return v.endsWith(':');
+    })
+  ) {
+    // TODO: add debug log
+    return {};
+  } else {
+    return tagMap;
+  }
+}
+
+function createOriginString(span: ReadableSpan): string | undefined {
+  // for some reason traceState keys must be w3c compliant and not stat with underscore
+  // using dd_origin for internal tracestate and setting datadog tag as _dd_origin
+  return (
+    span.spanContext.traceState &&
+    span.spanContext.traceState.get(OT_ALLOWED_DD_ORIGIN)
+  );
+}
+
+function createSpanName(span: ReadableSpan): string {
+  // span.instrumentationLibrary.name is not in v0.9.0 but has been merged
+  const instrumentationName = getInstrumentationName(span);
+  const spanKind = DD_SPAN_KIND_MAPPING[span.kind];
+
+  if (instrumentationName) {
+    const sanitizedName = instrumentationName
+      .replace('@', '')
+      .replace('/', '_');
+    return `${sanitizedName}.${spanKind}`;
+  } else if (spanKind) {
```

Comment:
```
if (instrumentationName) {
    const sanitizedName = instrumentationName
      .replace('@', '')
      .replace('/', '_');
    return `${sanitizedName}.${spanKind}`;
}
return spanKind || span.name;
```

Reply:
makes sense, appreciate the code snippet here, updated.